portfolio/src/main/webapp/pagination.js

```diff
@@ -0,0 +1,163 @@
+// Copyright 2019 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+document.addEventListener('DOMContentLoaded', () => {
+  const commentButton = new CommentButton();
+  commentButton.requestAndDisplayNewCommentsPage();
+});
+
+class CommentButton {
+  constructor() {
+    this.currentCommentsPageIndex = 0;
+    this.totalNumberOfPagesRequested = 1;
+    this.lastPage = -1;
+    this.cursorString = null;
+    this.prevCommentsButton = document.getElementById('prev-comments-button');
+    this.nextCommentsButton = document.getElementById('next-comments-button');
+    this.endOfCommentsSign = document.getElementById('end-of-comments-sign');
+    this.comments = document.getElementById('comments');
+
+    this.prevCommentsButton.classList.add("hidden");
+    this.endOfCommentsSign.classList.add("hidden");
+   
+    document.getElementById('maxNum').addEventListener('input', () => {
+      this.comments.innerHTML = "";
+      this.prevCommentsButton.classList.add("hidden");
+      this.nextCommentsButton.classList.remove("hidden");
+      this.endOfCommentsSign.classList.add("hidden");
+
+      this.currentCommentsPageIndex = 0;
+      this.totalNumberOfPagesRequested = 1;
+      this.lastPage = -1;
+      this.requestAndDisplayNewCommentsPage();
+    });
+
+    this.prevCommentsButton.addEventListener('click', () => {
+      this.displayPrevPage();
+    });
+
+    this.nextCommentsButton.addEventListener('click', () => {
+      this.displayNextPage();
+    });
+  }
+
+  hideCommentAtIndex(index) {
+    const commentList = this.comments.children;
+    commentList[index].classList.add("hidden");
+  }
+
+  showCommentAtIndex(index) {
+    const commentList = this.comments.children;
+    commentList[index].classList.remove("hidden");
+  }
+
+  displayNextPage() {
+    this.prevCommentsButton.classList.remove("hidden");
+
+    let commentList = this.comments.children;
+    if (this.currentCommentsPageIndex + 1 === this.lastPage) {
```

Comment:
This feels like a strange place to do this work. Shouldn't you set this after you load the last page OR at the bottom if the page you are going to ends up being the last page?

If you are in this situation, you are already requesting beyond the last page.

Reply:
The way that I have this.currentCommentsPageIndex set up makes it point to the page that is displayed before the user clicks on the next button. I added a clarification comment.

Reply:
Should you do an early return on line 88? 

if (this.currentCommentsPageIndex + 1 === this.lastPage) {
   // There are no more comments.
   update your hidden states
   return;
}